### ISSUE TITLE
Toward GeoTrellis 2.0

### ISSUE THREAD

**Issue by fosskers** [locationtech/geotrellis#2341]
<?xml version="1.0" encoding="utf-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN"
"http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en">
<head>
</head>
<body>
<div id="content">
<div id="table-of-contents">
<h2>Table of Contents</h2>
<div id="text-table-of-contents">
<ul>
<li><a href="#org226db47">1. <code>geotrellis-core</code></a>
<ul>
<li><a href="#org10ddcb8">1.1. Naming</a></li>
<li><a href="#org613a220">1.2. Importing</a></li>
<li><a href="#orgf19bb4b">1.3. Contents</a></li>
<li><a href="#org4a1bb4a">1.4. Migration</a></li>
</ul>
</li>
<li><a href="#org157664d">2. Discoverability</a>
<ul>
<li><a href="#org779e649">2.1. Reducing duplication</a></li>
<li><a href="#orgcd9e3b8">2.2. Avoiding method injection where deemed unnecessary</a></li>
</ul>
</li>
<li><a href="#org312f7f3">3. Easier SemVer iteration</a></li>
</ul>
</div>
</div>
<p>
Recent work with spin-off libraries, Raster Foundry, and GeoTrellis itself have
made us aware that we're nearing a "major version precipice". We've been holding back
a few PRs, knowing that they're necessary yet API breaking. And we've found a few quirks
with the project structure, the canonical one being:
</p>

<blockquote>
<p>
Why do I have to depend on <code>spark-core</code> just to get at <code>SpatialKey</code>?
</p>
</blockquote>

<p>
So, if we need to tighten our overalls, hit the keyboard, and break the API,
what's the best way to go about it? How can we get the most bang for our buck
if we're going to be incrementing our major version?
This doc serves as a launching point for that discussion and subsequent work.
</p>

<div id="outline-container-org226db47" class="outline-2">
<h2 id="org226db47"><span class="section-number-2">1</span> <code>geotrellis-core</code></h2>
<div class="outline-text-2" id="text-1">
<p>
@echeipesh  and I had a conversation in early May when he was trying to find a home
for <a href="https://github.com/locationtech/geotrellis/blob/25691da6fabee72c8dd8f9476af96ccdce8aa659/spark/src/main/scala/geotrellis/spark/buffer/Direction.scala#L21">a sum type</a> which didn't trivially belong to either <code>vector</code>, <code>raster</code>, or <code>spark</code>.
We thought it might be time to revive the idea of a separate <code>geotrellis-core</code> (or
likewise named) package which would always be at the bottom of any dependency
graph involving geotrellis.
</p>

<p>
More recently while working on a PR involving <a href="https://github.com/locationtech/geotrellis/pull/2334">MapKeyTransform</a>, certain intuitive
seeming transformation methods weren't possible to write given the way that the
various GT libraries depend on each other, for example:
</p>

<div class="org-src-container">
<pre class="src src-scala"><span style="color: #a45bad;">Extent</span>.toGridBounds<span style="color: #4f97d7; font-weight: bold;">:</span> <span style="color: #ce537a; font-weight: bold;">LayoutDefinition</span> <span style="color: #4f97d7; font-weight: bold;">=&gt;</span> <span style="color: #a45bad;">GridBounds</span>
</pre>
</div>

<p>
A contributor @metasim has <a href="https://github.com/locationtech/geotrellis/pull/2334#commitcomment-23745931">chimed in</a> with his support for this idea.
</p>

<p>
In general, a <code>geotrellis-core</code> package would be a place for the most central
types to live, regardless if the user is interested in vectors, rasters, vectortiles,
or Spark.
</p>
</div>

<div id="outline-container-org10ddcb8" class="outline-3">
<h3 id="org10ddcb8"><span class="section-number-3">1.1</span> Naming</h3>
<div class="outline-text-3" id="text-1-1">
<p>
Suggestions:
</p>

<ul class="org-ul">
<li><code>geotrellis-core</code> (Cats/ScalaZ/Circe/Monocle convention)</li>
<li><code>geotrellis-types</code> (Haskell convention)</li>
<li><code>geotrellis-common</code> (Java convention)</li>
</ul>
</div>
</div>

<div id="outline-container-org613a220" class="outline-3">
<h3 id="org613a220"><span class="section-number-3">1.2</span> Importing</h3>
<div class="outline-text-3" id="text-1-2">
<div class="org-src-container">
<pre class="src src-scala"><span style="color: #2aa1ae; background-color: #292e34;">/* </span><span style="color: #2aa1ae; background-color: #292e34;">To the point. Python/Cats/Circe/Haskell-like. Says "gimme it all, I trust you." */</span>
<span style="color: #4f97d7; font-weight: bold;">import</span> geotrellis.<span style="color: #4f97d7; font-weight: bold;">_</span>

<span style="color: #2aa1ae; background-color: #292e34;">/* </span><span style="color: #2aa1ae; background-color: #292e34;">True to the package name. */</span>
<span style="color: #4f97d7; font-weight: bold;">import</span> geotrellis.core.<span style="color: #4f97d7; font-weight: bold;">_</span>
</pre>
</div>
</div>
</div>

<div id="outline-container-orgf19bb4b" class="outline-3">
<h3 id="orgf19bb4b"><span class="section-number-3">1.3</span> Contents</h3>
<div class="outline-text-3" id="text-1-3">
<p>
I did a quick scan of the Scaladocs during my initial conversation with Eugene. I suggested
the following be moved to a central package:
</p>

<ul class="org-ul">
<li>Boundable</li>
<li>Bounds</li>
<li>EmptyBounds</li>
<li>KeyBounds</li>
<li>LayerId</li>
<li>Metadata</li>
<li>SpatialKey</li>
<li>SpaceTimeKey</li>
<li>TemporalKey</li>
<li>TileLayerMetadata</li>
<li>All CRSs? (from proj4)</li>
<li>All KeyIndex types and indexing functionality</li>
<li>LayoutDefinition and friends</li>
<li>Extent</li>
</ul>

<p>
I'm likely missing some. If you notice any, please let me know and I'll update this list.
Anything that can realistically be used regardless of vector/raster/spark could be here.
</p>
</div>
</div>

<div id="outline-container-org4a1bb4a" class="outline-3">
<h3 id="org4a1bb4a"><span class="section-number-3">1.4</span> Migration</h3>
<div class="outline-text-3" id="text-1-4">
<p>
If we adopt <code>import geotrellis._</code>, then the migration guide may amount to:
</p>

<blockquote>
<p>
Add <code>import geotrellis._</code> to any source file that no longer compiles. You may
be able to remove a number of imports that are no longer necessary.
</p>
</blockquote>
</div>
</div>
</div>

<div id="outline-container-org157664d" class="outline-2">
<h2 id="org157664d"><span class="section-number-2">2</span> Discoverability</h2>
<div class="outline-text-2" id="text-2">
<p>
This is an effort to address one of our team goals for the year - improving "discoverability"
in GeoTrellis.
</p>

<p>
We often get questions like the following from both within Azavea and without:
</p>

<blockquote>
<p>
How do I do &lt;common GIS op&gt; in GeoTrellis?
</p>
</blockquote>

<p>
or more frustrating:
</p>

<blockquote>
<p>
How do I import <code>Foo</code> in GeoTrellis?
</p>
</blockquote>

<p>
or the most frustrating:
</p>

<blockquote>
<p>
Q: I followed examples but my code just doesn't work. Help?
</p>

<p>
A: Probably missing implicits. Just import <code>geotrellis.foo.bar.baz._</code> and you're good to go.
</p>
</blockquote>

<p>
We can take a lesson from a library whose functionality is hard to discover and has nightmarish
imports: ScalaZ.
</p>

<p>
A suggested definition of <i>discoverable</i>:
</p>

<blockquote>
<p>
A library's API/functionality is <i>discoverable</i> if users can locate types, functions/methods,
and usage examples without asking humans for help and without looking at that library's code
directly.
</p>
</blockquote>

<p>
Assumptions:
</p>

<ul class="org-ul">
<li>Poor discoverability in Scala libraries is not inevitable.</li>
<li>An API with smaller "surface area" is more discoverable.</li>
<li>API breaks in pursuit of smaller surface area are okay.</li>
</ul>
</div>

<div id="outline-container-org779e649" class="outline-3">
<h3 id="org779e649"><span class="section-number-3">2.1</span> Reducing duplication</h3>
<div class="outline-text-3" id="text-2-1">
<p>
A number of our classes have heavily overloaded methods, where the inputs
vary only slightly. Many of these overloads perform simple <code>A -&gt; B</code> pure
transformations before eventually calling down to the "real" form of
the method. <a href="https://github.com/locationtech/geotrellis/blob/master/raster/src/main/scala/geotrellis/raster/render/ColorRamp.scala#L82">Here is <code>ColorRamp.toColorMap</code> as an example.</a> Which is the "real"
one?
</p>

<p>
Sets of overloads like these can make the call tree hard to follow for
developers. Since they're also often undocumented, the question can arise
in a user's mind, "what's the difference?" and they're forced to guess or ask us.
</p>

<p>
If these assumptions are true, would it be prudent to reduce the number of such
overloads, putting the burden of the <code>A -&gt; B</code> transformation onto the user?
Exceptions can be made of course, one coming to mind being one of the ways
to construct a Polygon:
</p>

<div class="org-src-container">
<pre class="src src-scala"><span style="color: #a45bad;">Polygon</span><span style="color: #4f97d7;">(</span><span style="color: #bc6ec5;">(</span><span style="color: #a45bad;">0</span>,<span style="color: #a45bad;">0</span><span style="color: #bc6ec5;">)</span>, <span style="color: #bc6ec5;">(</span><span style="color: #a45bad;">0</span>,<span style="color: #a45bad;">1</span><span style="color: #bc6ec5;">)</span>, <span style="color: #bc6ec5;">(</span><span style="color: #a45bad;">1</span>,<span style="color: #a45bad;">1</span><span style="color: #bc6ec5;">)</span>, <span style="color: #bc6ec5;">(</span><span style="color: #a45bad;">1</span>,<span style="color: #a45bad;">0</span><span style="color: #bc6ec5;">)</span>, <span style="color: #bc6ec5;">(</span><span style="color: #a45bad;">0</span>,<span style="color: #a45bad;">0</span><span style="color: #bc6ec5;">)</span><span style="color: #4f97d7;">)</span>
</pre>
</div>

<p>
Having to prefix each point with a <code>Point</code> constructor would get verbose.
Then again, how often do we manually create Polygons like this, outside of tests?
</p>

<p>
Here are a few candidates for reducing duplicates.
</p>

<ul class="org-ul">
<li><code>PngRenderMethods.renderPng</code></li>
<li><code>MapKeyTransform</code> methods</li>
<li><code>ColorRamp.toColorMap</code></li>
</ul>
</div>
</div>

<div id="outline-container-orgcd9e3b8" class="outline-3">
<h3 id="orgcd9e3b8"><span class="section-number-3">2.2</span> Avoiding method injection where deemed unnecessary</h3>
<div class="outline-text-3" id="text-2-2">
<p>
We have a vast, undocumented mechanism for method injection involving 411 classes/traits. They follow
the naming pattern <code>*Methods</code> and <code>with*Methods</code>. They can be considered a source of confusion,
as they add a layer of indirection in determining the location of method definitions.
This leads to confusion with imports, and slows down development.
</p>

<p>
A number of these sets of methods (likely not all) could live directly in the
class they operate on. This would reduce a huge swath of symbols that appear
in Scala docs, and would likely decrease compile time.
</p>

<p>
Examples:
</p>

<ul class="org-ul">
<li><code>ColorMethods</code></li>
</ul>
</div>
</div>
</div>

<div id="outline-container-org312f7f3" class="outline-2">
<h2 id="org312f7f3"><span class="section-number-2">3</span> Easier SemVer iteration</h2>
<div class="outline-text-2" id="text-3">
<p>
If every symbol available in GeoTrellis is visible to the user through imports or
documentation, then every symbol is "user facing", and every change has the potential
to be API breaking. This makes it particularly hard to make sound calls regarding our versioning.
</p>

<p>
If we make a sweep of GT and draw a clear line in the sand between what is user-facing
and what isn't, then whatever isn't can be hidden via <code>private</code> or stuffed in an
<code>internal</code> subpackage (a Haskell convention that works well that I use in both VT and VP).
</p>

<p>
This should make it easier to do heavier changes without causing API breaks. This would
allow us to increment versions easier and keep a regular release cycle. This also has
the benefit of reducing API surface area, since many symbols will be made hidden from
the Scaladocs.
</p>
</div>
</div>
</div>

</body>
</html>

@metasim

**Comment by fosskers** [comment]
There's a code example there for imports which didn't seem to survive the conversion from `.org`. Here is what I intended:

```scala
  /* To the point. Python/Cats/Circe/Haskell-like. Says "gimme it all, I trust you." */
  import geotrellis._

  /* True to the package name. */
  import geotrellis.core._
```

**Comment by metasim** [comment]
I raise my hand for `geotrellis-core` over `geotrellis-types`, as it leaves the cognitive door open for content that aren't just ADTs or are are commonly used pure functions.

**Comment by fosskers** [comment]
All typeclass instances for those types should also live in `geotrellis-core`. In GT's case this mainly means the JSON codecs.

Crap, speaking of which, I forgot to mention JSON. Maybe Circe migration should be 3.0?

**Comment by metasim** [comment]
Avro codecs too?

**Comment by lossyrob** [comment]
> Boundable
> Bounds
> EmptyBounds
>KeyBounds
> LayerId
> Metadata
> SpatialKey
> SpaceTimeKey
> TemporalKey
> TileLayerMetadata
> LayoutDefinition and friends

All this has to do with tiling and IO. Seems more like this should be broken out into an core `io` module, or something that has to do with tiling. 

> All CRSs? (from proj4)

We already have a subproject that has all the CRS's - proj4. There's also an issue to handle other projection libraries (https://github.com/locationtech/geotrellis/issues/1294). I was imagining the refactor here being we have a `geotrellis.crs` library, that can detect if another library is present and use it (e.g. using proj4j, proj.4 bindings or Apache SIS).

> All KeyIndex types and indexing functionality

When we move to supporting SFCurve, this will carry an additional dependency, and be worth a project in it's own right (e.g. geotrellis-indexing)

> Extent

This is a vector type, I think it belongs in geotrellis-vector.

If we carry through all the JSON and Avro, we adding a lot of dependencies to things that don't need it (e.g. raster doesn't need avro, and people user raster on it's own).

I think some sort of different breakdown is definitely needed. I'm not sure this `core` vs `non-core` representation is it.

**Comment by fosskers** [comment]
> Seems more like this should be broken out into an core io module

A quote from @echeipesh from the original May conversation: 

> Ok, I think there is a set of functionality, working with tiled layouts. That is a common pattern for what we use, but it doesn’t actually have a lot to do with raster.

I agree that those types have a lot to do with tiling, but I'm not seeing the connection to `IO`. Surely those things are *used* by our IO backends (`s3`, etc), but the types themselves are GIS concepts that can be used in pure code that doesn't involve spark or the idea of data passing through a program. Related to that, it's always struct me as odd that our GeoJSON typeclass instances live in `vector.io`. Converting between JSON and data types in and of itself has nothing to do with IO, unless we're not sharing the same definition of IO?

> We already have a subproject that has all the CRS's - proj4.

```scala
lazy val proj4 = project
  .settings(commonSettings)
  .settings(javacOptions ++= Seq("-encoding", "UTF-8"))
```
Cool, looks like `proj4` is at the bottom of the dep graph too. Might not be a problem for it to stay that way, which means that `geotrellis-core` might even be able to depend on it.

> When we move to supporting SFCurve, this will carry an additional dependency, and be worth a project in it's own right (e.g. geotrellis-indexing)

:+1: 

> [Extent] is a vector type, I think it belongs in geotrellis-vector.

@echeipesh said the same thing, but it's *damn* central to tiling work. I realize that there's a close relationship between `Extent` and `Polygon`, but `Extent` by itself isn't unique to working with geometries.  I'd say it's as fundamental to any GeoTrellis work at `SpatialKey` is.

**Comment by echeipesh** [comment]
Overall agree with need to simplify the import burden for using GeoTrellis.

## `-core` vs `-tiling` breakout

Otherwise the main drivers for the class shuffle have been the need to use Collections API without Spark dependency. In general this leads down the road of working with tiled rasters somewhere other than `geotrellis-spark`. The includes

- Reading tiled values/rasters
- Writing tiled values/rasters
- Operations over tiled values/rasters

All of these things can live pretty happily in `geotrellis-raster` but that does scope their re-use somewhat. In fact there is a lot of functionality that has to do with laying out either pixels or tiles over map surface and a lot of the classes @fosskers mentioned are the bottom of that class hierarchy. With exception of `LayerId` none of them have anything specifically to do with IO, thats just the place where they are most prominently used. The gut feeling is that it makes sense to home those in `geotrellis-core` rather than trying to find a very specific package name.

## Codecs

Another pain felt is the lack of a good home for Avro/JSON codecs. The pain compounds when you consider library choices. Migrating to `circe` would be nice but cutting people loose from `spray-json` could also be problematic as it would force to all downstream applications to deal with swapping in circe which as we saw with Raster Foundry comes with its own gotchas. Similar story goes for Avro if you consider that ProtoBuf is at least one more valid choice.

The answer hopefully is to separate the GeoTrellis interfaces for specific library kind of like its done with `Component`, abstraction over lenses, and then break out the Avro and spray-json into their own packages.

**Comment by fosskers** [comment]
Right, duh, the collections api is a much stronger motivation for `geotrellis-core` than my `SpatialKey` woes. ;p

### Orphan Instances

Re: where should json instances / avro codecs go?

We implement typeclasses in Scala with implicits. These are sometimes implicit `class` wrappings, and sometimes `implicit val`s  which provide instances of classes needed by some external library functionality (like `Encoder` from Circe). If these implicits are defined anywhere but the companion object of the type they're tied to, they become "orphans".

Said another way, using `Encoder` as an example: if it is possible to import a type `A` while *not* importing its `Encoder` instance, then that instance is an orphan.

**Claim:** Orphans increase user cognitive overhead, specifically regarding imports.

[Example from our FAQ regarding `.toGeoJson`](https://geotrellis.readthedocs.io/en/latest/guide/faq.html#how-do-i-import-geotrellis-methods).

Defining orphans leads to compiler warnings in languages with first-class typeclasses, so in Scala it's just something we have to be proactive about. 

That said, does anyone foresee problems with migrating json / avro codecs to the companion objects of their types? This is how Circe instance auto-generation works - the instances are written into companion objects. The library breakout approach suggested by @echeipesh in his Codecs section above has value to be sure, although the fact remains that the instances themselves would be orphans.

**Comment by moradology** [comment]
As an alternative to stuffing would-be orphans into companions, we could have spray and circe alternative imports and stuff all required codecs into a single import object. Barring that, maybe we could also wrap the spray encoders/decoders we have with circe (did this a bit in RFML) to give nice operation within a `circe` application

**Comment by fosskers** [comment]
I just hit this again, motivating for us to attempt to avoid orphans. This won't compile:

```scala
import geotrellis.raster._
import geotrellis.spark._
import geotrellis.spark.io.s3._
import org.apache.spark._

// --- //

object Pipeline extends App {
  override def main(args: Array[String]): Unit = {
    implicit val sc: SparkContext = new SparkContext(
      new SparkConf().setMaster("local[*]").setAppName("pipeline-test")
    )

    val reader = S3LayerReader("geotrellis-test", "colingw/pipeline")
    val layer = LayerId("mask", 14)

    val rdd = reader.read[SpatialKey, Tile, TileLayerMetadata[SpatialKey]](layer)

    println(s"COUNT: ${rdd.count}")

    sc.stop()
    println("done")
  }
}
```
saying
```
could not find implicit value for evidence parameter of type spray.json.JsonFormat[geotrellis.spark.SpatialKey]
```
So having `SpatialKey` in scope is not enough to read a layer that uses them. Trial and error reveals this:
```scala
import geotrellis.spark.io._
```
to be the necessary extra import.

**Comment by fosskers** [comment]
An idea from @lossyrob yesterday, for some future version of GeoTrellis. Paraphrased:

> We shouldn't wrap the JTS geometry types ourselves. All our functions should work on JTS geoms directly, with extra functionality added by injected methods only when necessary.

**Comment by fosskers** [comment]
^ Although it occurs to me now that this would force the user to import (and depend on!) JTS themselves when they want to reference the Geom types.

Workaround until Scala supports re-exports?

```scala
import blah.blah.jts

package object vector {
  type Point = jts.Point
  ...  // same for other geom types
}
```

**Comment by metasim** [comment]
@fosskers From where I stand I agree wholeheartedly with @lossyrob and really like the idea of using type aliases to simplify imports. Sounds like a winning approach to me.
